adding an option for the filter ipsubnet and testing

pierremahot opened the pull request:
##### SUMMARY
Adding the option to request the index of a subnet in a bigger subnet

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
filter ipsubnet

##### ANSIBLE VERSION
```
ansible 2.5.2
  config file = ansible.cfg
  python version = 2.7.9 default, jun 29 2016, 13:08:31) [GCC 4.9.2]
```
##### ADDITIONAL INFORMATION
Need this feature to deduce easily a vlan for my switch
```
Examples:
---
- hosts: local
  tasks:
    - debug: msg="{{ '192.168.50.0/24' | ipsubnet('192.168.0.0/16') }}"
#display 51
    - debug: msg="{{ '10.10.10.8/30' | ipsubnet('192.168.0.0/16')}}
#display False
   - debug: msg="{{ '10.10.10.8/16' | ipsubnet('10.10.0.99/24')}}
#display False
   - debug: msg="{{ '10.10.0.9/30' | ipsubnet('10.10.0.99/24')}}
#display 3
```

pierremahot committed:
adding an option for the filter ipsubnet

adding an option for the filter ipsubnet

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -718,7 +718,11 @@ def ipv6(value, query=''):
+#  - address/prefix | ipsubnet(subnet/prefix)
+#      return the index of the subnet in the subnet
 def ipsubnet(value, query='', index='x'):
+    strquery = str(query)
```
Please don't add anything before a docstring.

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -734,8 +738,7 @@ def ipsubnet(value, query='', index='x'):
 
     if not query:
```
better add it above this line

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/67119/3/tests

```
>       self.assertEqual(ipsubnet(address, subnet), False)
E       AssertionError: '35' != False

test/units/plugins/filter/test_ipaddr.py:490: AssertionError
```

pierremahot committed:
correct tests and do str(query) only once

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -735,10 +734,11 @@ def ipsubnet(value, query='', index='x'):
     except:
         return False
-
+    querystr = str(query)
```
Also, since in Python we use `snake_case` for variables/functions, please change this to `query_string`

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -767,6 +770,23 @@ def ipsubnet(value, query='', index='x'):
+                v = ipaddr(query, 'subnet')
+
+            query = netaddr.IPNetwork(v)
+        except:
```
It is a bad practice to catch all exceptions wholesale. You mask different errors, which you should've processed correctly. Also this makes it almost impossible to debug misbehavior.

pierremahot replied:
I just make the same code as before line 727 to 736 but I am not sure how to cut this the good way : 
```python
        try:
            vtype = ipaddr(query, 'type')
       except:
            return False

        if vtype == 'address':
            try:
                v = ipaddr(query, 'cidr')
            except:
                return False
        elif vtype == 'network':
            try:
                v = ipaddr(query, 'subnet')
            except:
                return False
        query = netaddr.IPNetwork(v)
```

webknjaz replied:
Yes, I see that, but it's not a good approach.
From https://github.com/drkjam/netaddr/blob/ebf714601e8cd6234575422f876d4d579b202b1a/netaddr/ip/__init__.py#L938 it looks like `netaddr.IPNetwork(v)` may raise `AddrFormatError`.

But if you want to catch all `netaddr` library's thrown exceptions, you may do:
```python
from netaddr.core import AddrConversionError, AddrFormatError, NotRegisteredError
NETADDR_ERRORS = AddrConversionError, AddrFormatError, NotRegisteredError
...
try:
    ...your exceptional code here
except NETADDR_ERRORS as err:
    process_or_log(err)
    # optionally raise/pass/return False (whatever fits)
```

webknjaz replied:
I still insist on correct exception handling.

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -767,6 +770,23 @@ def ipsubnet(value, query='', index='x'):
+        if network_in_network(query, value):
+            subnetlist = list(query.subnet(value.prefixlen))
+            for i in range(0, len(subnetlist)):
+                if (subnetlist[i] == value):
```
In Python, we don't wrap clauses with parentheses. Please remove them.

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -767,6 +770,23 @@ def ipsubnet(value, query='', index='x'):
+
+        if network_in_network(query, value):
+            subnetlist = list(query.subnet(value.prefixlen))
+            for i in range(0, len(subnetlist)):
```
What you're trying to do is possible to achieve using "sliding window" technique with `enumerate()`, it will be possible to even drop dancing around "index+1" problem:
```python
for i, subnet in enumerate(query.subnet(value.prefixlen)):
    if subnet == value:
        return str(i+1)
```

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -767,6 +770,23 @@ def ipsubnet(value, query='', index='x'):
+            return False
+
+        if network_in_network(query, value):
+            subnetlist = list(query.subnet(value.prefixlen))
```
According to the doc, it's a generator: https://netaddr.readthedocs.io/en/latest/api.html#netaddr.IPNetwork.subnet

So you can iterate over it, no need to `list()` it. Please remove it.

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -767,6 +770,23 @@ def ipsubnet(value, query='', index='x'):
+        except:
+            return False
+
+        if network_in_network(query, value):
```
I believe, you don't even need the if-wrapper, if `query.subnet` won't yield anything, the loop will be "skipped", just like your intention is.

pierremahot committed:
make the @webknjaz recommandation

change the for loop to iteration, delete the if statement useless and change the variable strquery to query_string

pierremahot committed:
correction of the enumerate start

pierremahot commented:
It's my choice i prefert to have the first subnet to have the value 1 and not 0

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -782,7 +782,7 @@ def ipsubnet(value, query='', index='x'):
             return False
 
-        for i, subnet in enumerate(query.subnet(value.prefixlen)):
+        for i, subnet in enumerate(query.subnet(value.prefixlen), 1):
```
:heart_eyes: TIL

webknjaz commented:
I've restarted failed CI jobs.

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -473,3 +473,32 @@ def test_cidr_merge(self):
         self.assertEqual(cidr_merge(subnets), ['1.12.1.1/32', '1.12.1.255/32'])
         self.assertEqual(cidr_merge(subnets, 'span'), '1.12.1.0/24')
+
+    def test_ipsubnet(self):
```
where's negative tests?

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -473,3 +473,32 @@ def test_cidr_merge(self):
+        self.assertEqual(ipsubnet(address, subnet), False)
+        address = '192.168.144.5'
+        subnet = '192.168.0.0/16'
+        self.assertEqual(ipsubnet(address), '192.168.144.5/32')
```
Please keep your code style consistent: first you assign vars for some time, then you pass lots of function args directly. There's a better way.

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -473,3 +473,32 @@ def test_cidr_merge(self):
+
+    def test_ipsubnet(self):
+        address = '1.1.1.1/24'
+        self.assertEqual(ipsubnet(address, '30'), '64')
```
You could create a helper method:
```python
def _test_ipsubnet(self, ipsubnet_args, expected_result):
    self.assertEqual(ipsubnet(*ipsubnet_args), expected_result)

# and then reuse it by looping over test data:
def test_ipsubnet(self):
    test_cases = (
        (('1.1.1.1/25', ), '0'),
        ...
        (('192.168.144.5', '18', '-5'), '192.168.144.0/27'),
    )
    for args, res in test_cases:
        self._test_ipsubnet(args, res)
```

pierremahot committed:
update testing and adding negative tests

pierremahot committed:
missing line at end of file

gundalow commented:
@pierremahot Can you please rebase and resolve the merge conflict so we can look at getting this in, thanks https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html

pierremahot committed:
merge

pierremahot committed:
reabasing

pierremahot committed:
reabasing

pierremahot committed:
reabasing

pierremahot committed:
reabasing

pierremahot committed:
import ipsubnet after rebase

pierremahot committed:
forgeted comma

pierremahot committed:
update test false input

pierremahot committed:
rmake the validated test

gundalow commented:
Thanks for the rebase, could you please update the exception handling as suggested by webknjaz.

Also we need documentation for this, which goes in `docs/docsite/rst/user_guide/playbooks_filters_ipaddr.rst`
Thanks in advance.

pierremahot committed:
update documentation, adding documentation exemple to thet tests and make the exption as webknjaz proposed

pierremahot committed:
remove useless space and move import to better place

webknjaz commented on docs/docsite/rst/user_guide/playbooks_filters_ipaddr.rst:
```diff
@@ -474,6 +474,28 @@ smaller subnets::
+    36870
+
+    # The rank in the /24 that contain the address
+    # {{ address | ipsubnet(192.168.144.0/24) }}
```
I bet that the argument should be wrapped in quotes (I might be wrong though)

webknjaz replied:
@pierremahot did you test this?

pierremahot replied:
You are right the quote are needed.

pierremahot replied:
I have Tested it in a playbook with the quote it's ok 
test.yml file :
```yml
- hosts: localhost
    tasks:
        - debug:
            msg: "{{ '1.12.1.34/32' | ipsubnet('1.12.1.34/24') }}"
```
and the result :
```
ansible-playbook test.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] *****************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "35"
}

PLAY RECAP ***********************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0

```

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -24,6 +24,8 @@
 try:
     import netaddr
+    from netaddr.core import AddrConversionError, AddrFormatError, NotRegisteredError
+    NETADDR_ERRORS = AddrConversionError, AddrFormatError, NotRegisteredError
```
I don't like having this variable in a single code path. You should have it defined at all times.

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -24,6 +24,8 @@
 
```
Please add the following here:
```python
NETADDR_ERRORS = ()
```

webknjaz reviewed (changes requested):
Please make the code more stable by following my comments below.

pierremahot committed:
adding NETADDR_ERRORS variable all times

pierremahot commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -29,6 +29,7 @@
 except ImportError:
     # in this case, we'll make the filters return error messages (see bottom)
     netaddr = None
+    NETADDR_ERRORS = ()
```
I put it here like netaddr don't know if it's more logic

webknjaz replied:
Yes, you can do it like this.

pierremahot committed:
adding quote

webknjaz commented on lib/ansible/plugins/filter/ipaddr.py:
```diff
@@ -785,6 +791,19 @@ def ipsubnet(value, query='', index='x'):
+                v = ipaddr(query, 'subnet')
+            query = netaddr.IPNetwork(v)
+        except NETADDR_ERRORS as err:
+            raise errors.AnsibleFilterError(err)
```
This is minor, but I like to keep exception chain when re-raising new ones. This is how you do it in forward-compatible manner:
```python
from ansible.module_utils import six
...
	except NETADDR_ERRORS as err:
            six.raise_from(errors.AnsibleFilterError(err), err)
```

pierremahot committed:
add re-raising exception chain as webknjaz tell

webknjaz reviewed (approved):
:+1: :tada:

pierremahot commented:
I have a problème with the exception :( doesn't look to work it return 1

```yml
- hosts: localhost
  tasks:
    - debug:
        msg: "{{ '192.168.1.1/24' | ipsubnet('-5') }}"
```

result
```
ansible-playbook test.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] *****************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "1"
}

PLAY RECAP ***********************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

webknjaz commented:
@pierremahot can you add that to integration tests?

webknjaz commented:
Also, what do you mean by exception?

pierremahot committed:
adding exception test

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -502,3 +502,40 @@ def test_ipmath(self):
+            self._test_ipsubnet(args, res)
+
+    def _test_ipsubnet(self, ipsubnet_args, expected_result):
+        self.assertEqual(ipsubnet(*ipsubnet_args), expected_result)
```
I'd also add some test cases with assertRaises

pierremahot replied:
I think 
```
vtype = ipaddr(query, 'type')
        if vtype == 'address':
            v = ipaddr(query, 'cidr')
        elif vtype == 'network':
            v = ipaddr(query, 'subnet')
        else:
            return False
```
Can't raise error on : 
```
query = netaddr.IPNetwork(v)
for i, subnet in enumerate(query.subnet(value.prefixlen), 1)
```

pierremahot replied:
i don't find the case

webknjaz replied:
You could `mock.patch` it to do so :)

webknjaz replied:
Also, look out for netaddr docs for the cases when it'll raise the exception.

pierremahot committed:
testing assertion

pierremahot commented:
I can't have assertion error because of the way ipaddr works

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -541,3 +539,13 @@ def test_ipsubnet(self):
+        expected = 'An Exception indicating a failure to convert between address types or notations.'
+        with self.assertRaises(AddrConversionError) as exc:
+            ipsubnet('192.168.144.5', '192.168.8.1.5')
+            self.assertEqual(exc.exception.message, expected)
```
This needs to go outside of CM. If `ipsubnet` will raise exception it will interrupt code block within `with`.

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -502,3 +502,50 @@ def test_ipmath(self):
+        self.assertEqual(ipsubnet(*ipsubnet_args), expected_result)
+
+        expected = 'An Exception indicating a network address is not correctly formatted.'
+        with self.assertRaises(AddrConversionError) as exc:
```
I think, you're looking for [`self.assertRaisesRegexp(AddrConversionError, expected)`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertRaisesRegexp)

pierremahot replied:
I don't find case to raise an error

pierremahot committed:
no raise possible

pierremahot committed:
update test and code

pierremahot commented:
I think ipaddr is the function that can raise error but it filter already the input of netaddr i don't find case where it raise an error

webknjaz commented:
Yeah, you're right: `ipaddr` brutally swallows all exceptions.

pierremahot committed:
raise error when out of subnet or wrong subnet input

pierremahot committed:
update doc for raising error

webknjaz commented on test/units/plugins/filter/test_ipaddr.py:
```diff
@@ -541,3 +537,12 @@ def test_ipsubnet(self):
     def _test_ipsubnet(self, ipsubnet_args, expected_result):
         self.assertEqual(ipsubnet(*ipsubnet_args), expected_result)
+        expected = 'You must pass a valid subnet or IP address; invalid_subnet is invalid'
+        with self.assertRaises(AnsibleFilterError) as exc:
```
Please use `assertRaisesRegexp` instead.

pierremahot committed:
use asserRaisesRegexp
